[iOS] Pan gestures immediately after tap do not scroll a page with an unresponsive viewport

whsieh opened the pull request:
#### f9672dbf09d92cbcc1ce47ee959a7ff55ed27f9e
<pre>
[iOS] Pan gestures immediately after tap do not scroll a page with an unresponsive viewport
<a href="https://bugs.webkit.org/show_bug.cgi?id=255300">https://bugs.webkit.org/show_bug.cgi?id=255300</a>
rdar://106010038

Reviewed by Megan Gardner.

Currently, in a page with an unresponsive viewport, attempting to scroll the web view with a pan
gesture immediately after performing a tap does not actually result in any scrolling, since the
scroll view&apos;s pan gesture recognizer does not begin. This happens because while the pan gesture is
immediately reset to `Possible` state after the tap, it then transitions to `Failed` state right
away when subsequently panning the web view, since it&apos;s excluded by the synthetic tap (single click)
gesture which is still in `Ended` state because it hasn&apos;t been reset yet (due to the double click
gesture still being `Possible`).

To avoid this, we allow both the synthetic tap gesture to recognize simultaneously alongside pan
gestures by adding logic to `-gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:`.
This means that instead of transitioning to `Failed` state right away when panning the web view, we
instead remain in `Possible` state, which then allows us to recognize the gesture if the user begins
to actually pan the view.

Test: fast/events/touch/ios/swipe-after-tap-scrolls-web-view-in-non-responsive-viewport.html

* LayoutTests/fast/events/touch/ios/swipe-after-tap-scrolls-web-view-in-non-responsive-viewport-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/swipe-after-tap-scrolls-web-view-in-non-responsive-viewport.html: Added.

Add a layout test to exercise the fix. This test preemptively tries to mitigate any potential
flakiness as a result of either the tap or swipe gesture not getting recognized by automatically
retrying itself several times, until it&apos;s successful.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:]):

Canonical link: <a href="https://commits.webkit.org/262853@main">https://commits.webkit.org/262853@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da4d5052ba2169cca1cee22d99c1d639f72f6628

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2923 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4189 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3130 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2919 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2874 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2444 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2803 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3224 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2481 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3962 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/718 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2465 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2328 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2475 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2516 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3711 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2852 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2287 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2511 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2468 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/693 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2503 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2667 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

megangardner commented on LayoutTests/fast/events/touch/ios/swipe-after-tap-scrolls-web-view-in-non-responsive-viewport.html:
```diff
@@ -0,0 +1,65 @@
+
+    const maximumNumberOfAttempts = 10;
+    for (let i = 0; i < maximumNumberOfAttempts; ++i) {
+        if (await attemptTest()) {
```
Why attempt more than once?

whsieh replied:
I think there've been previous instances where simulated swipe gestures like this are flaky in automation; especially in a test as intricate as this one, baking retries into the test itself should help mitigate any potential flakiness.

pxlcoder replied:
Would you mind adding that explanation to the commit message?

whsieh replied:
Yep — will add blurb explaining that.

whsieh committed:
[iOS] Pan gestures immediately after tap do not scroll a page with an unresponsive viewport
https://bugs.webkit.org/show_bug.cgi?id=255300
rdar://106010038

Reviewed by Megan Gardner.

Currently, in a page with an unresponsive viewport, attempting to scroll the web view with a pan
gesture immediately after performing a tap does not actually result in any scrolling, since the
scroll view's pan gesture recognizer does not begin. This happens because while the pan gesture is
immediately reset to `Possible` state after the tap, it then transitions to `Failed` state right
away when subsequently panning the web view, since it's excluded by the synthetic tap (single click)
gesture which is still in `Ended` state because it hasn't been reset yet (due to the double click
gesture still being `Possible`).

To avoid this, we allow both the synthetic tap gesture to recognize simultaneously alongside pan
gestures by adding logic to `-gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:`.
This means that instead of transitioning to `Failed` state right away when panning the web view, we
instead remain in `Possible` state, which then allows us to recognize the gesture if the user begins
to actually pan the view.

Test: fast/events/touch/ios/swipe-after-tap-scrolls-web-view-in-non-responsive-viewport.html

* LayoutTests/fast/events/touch/ios/swipe-after-tap-scrolls-web-view-in-non-responsive-viewport-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/swipe-after-tap-scrolls-web-view-in-non-responsive-viewport.html: Added.

Add a layout test to exercise the fix. This test preemptively tries to mitigate any potential
flakiness as a result of either the tap or swipe gesture not getting recognized by automatically
retrying itself several times, until it's successful.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:]):

Canonical link: https://commits.webkit.org/262853@main
